### PR TITLE
Assignment 1 - Indigo Gift Voucher Assignment for Web

### PR DESCRIPTION
**Pull request by AmanDaharwal**
PR contains -
Feature file for Indigo Gift Voucher scenario ( As a guest user, Personalise and preview an Indigo gift voucher with invalid Promo code) 
Step class file for feature
BL class for feature
Abstract Screen classes for BL and screen class for web implementation
testdata.json and config.json file updates as per feature

### CODE REVIEW

File: src/test/resources/testData.json [hunk 1]
```diff
@@ -7,6 +7,11 @@
       "password": "eotTesting@0",
       "firstName": "Eot",
       "lastName": "Testing"
+    },
+    "aman.daharwal": {
```
Comment (anandbagmar) [inline src/test/resources/testData.json:11]: This is incorrect. 
The data is per environment. If you need to create user-specific data, then 
1. you need to add it for every environment
2. you need to update existing data to also fit in that pattern

File: src/test/resources/com/znsio/sample/e2e/features/indigoFlightPromoCode.feature [hunk 2]
```diff
@@ -0,0 +1,9 @@
+
+  @indigo
+Feature: Indigo gift voucher with invalid promo code
+
+  @web @android
+Scenario: As a guest user, Personalise and preview an Indigo gift voucher with invalid Promo code
+  Given I as a guest user, preview and personalise "2" gift vouchers of "5000" rupees
+  When I apply a "invalid" promo code
+  Then I should able to purchase the ticket of "10000" rupees
```
Comment (anandbagmar) [inline src/test/resources/com/znsio/sample/e2e/features/indigoFlightPromoCode.feature:9]: You are not purchasing a ticket. You are purchasing a gift voucher

Also, mentioning 10000 here means if you change the data in the Given step, you need to also update the data in the Then step.
I feel you can avoid mentioning 10000 in the Then step. Instead you could say "Then I should be able to purchase the gift voucher at the original price"

File: src/test/java/com/znsio/sample/e2e/steps/IndigoFlightSteps.java [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+    public void iAsAGuestUserPreviewAndPersonaliseTheGiftVouchersOfRupees(String quantity, String denominations) {
+        LOGGER.info("Preview and Personalise Gift Voucher");
+        allDrivers.createDriverFor(SAMPLE_TEST_CONTEXT.ME, Runner.platform, context);
+        new GiftVoucherBL().previewAndPersonaliseGiftVoucher(quantity, denominations);
+    }
+
+    @When("I apply a {string} promo code")
+    public void iApplyAPromoCode(String promoCode) {
+        LOGGER.info("Apply Promo code with delivery details");
+        new GiftVoucherBL().applyAPromoCode(promoCode).withDeliveryOptions();
```
Comment (anandbagmar) [inline src/test/java/com/znsio/sample/e2e/steps/IndigoFlightSteps.java:45]: call 1 BL method from the step. 
withDeliveryOptions is not a BL method

File: src/test/java/com/znsio/sample/e2e/steps/IndigoFlightSteps.java [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+        LOGGER.info("context: " + context.getTestName());
+        allDrivers = (Drivers) context.getTestState(SAMPLE_TEST_CONTEXT.ALL_DRIVERS);
+        LOGGER.info("allDrivers: " + (null == allDrivers));
+    }
+
+    @Then("I should able to purchase the ticket of {string} rupees")
+    public void iShouldAbleToPurchaseTheTicketOfRupees(String amount) {
+
+        LOGGER.info("Process Payment");
+        new GiftVoucherBL().processPayment(amount);
```
Comment (anandbagmar) [inline src/test/java/com/znsio/sample/e2e/steps/IndigoFlightSteps.java:32]: method name should be purchaseGiftVoucher, not processPayment

File: src/test/java/com/znsio/sample/e2e/screen/indigo/web/PaymentScreenWeb.java [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+        this.driver = driver;
+        this.visually = visually;
+        innerDriver = this.driver.getInnerDriver();
+    }
+
+    @Override
+    public int getPaymentAmount() {
+
+        String amount =  driver.waitTillElementIsPresent(byPaymentAmountXpath).getText();
+        return Integer.parseInt(amount.substring(4, amount.length() - 3));
```
Comment (anandbagmar) [inline src/test/java/com/znsio/sample/e2e/screen/indigo/web/PaymentScreenWeb.java:29]: will be good to add a comment here with an example why you are having this substring logic

File: src/test/java/com/znsio/sample/e2e/screen/indigo/web/GiftVoucherScreenWeb.java [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+
+        return this;
+    }
+
+    @Override
+    public GiftVoucherScreen personaliseGiftVoucher(Map senderDetails) {
+
+        LOGGER.info("Personalise gift voucher with details "+senderDetails.toString());
+        WebElement makeItPersonalWebElement = driver.waitTillElementIsPresent(byMakeItPersonalId);
+        if(!makeItPersonalWebElement.isSelected())
```
Comment (anandbagmar) [inline src/test/java/com/znsio/sample/e2e/screen/indigo/web/GiftVoucherScreenWeb.java:92]: add {}

File: src/test/java/com/znsio/sample/e2e/screen/indigo/web/GiftVoucherScreenWeb.java [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+import org.apache.commons.lang.NotImplementedException;
+import org.apache.log4j.Logger;
+import org.openqa.selenium.By;
+import org.openqa.selenium.WebDriver;
+import org.openqa.selenium.WebElement;
+import org.openqa.selenium.support.ui.Select;
+
+import java.util.Map;
+
+public class GiftVoucherScreenWeb extends GiftVoucherScreen {
```
Comment (anandbagmar) [inline src/test/java/com/znsio/sample/e2e/screen/indigo/web/GiftVoucherScreenWeb.java:17]: You are not doing any visual validations anywhere in any of the screens. Update appropriately

File: src/test/java/com/znsio/sample/e2e/businessLayer/indigo/GiftVoucherBL.java [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+        LOGGER.info("Select delivery options - "+deliveryOption);
+
+        GiftVoucherScreen.get().selectDeliveryOption(deliveryOption)
+                .enterSenderDetails(details)
+                .checkTermsAndConditions(true);
+
+        return this;
+    }
+
+    public GiftVoucherBL processPayment(String amount) {
```
Comment (anandbagmar) [inline src/test/java/com/znsio/sample/e2e/businessLayer/indigo/GiftVoucherBL.java:102]: you are not processing the payment, just checking the payment amount.

File: src/test/java/com/znsio/sample/e2e/businessLayer/indigo/GiftVoucherBL.java [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+        String promo = "";
+        if(promoCode.equals("invalid")){
+            promo = Randomizer.randomizeAlphaNumericString(8);
+        }
+        else{
+            promo = promoCode;
+        }
+        GiftVoucherScreen.get().applyPromoCode(promo);
+
+        assertThat(GiftVoucherScreen.get().getPromoCodeMessage())
```
Comment (anandbagmar) [inline src/test/java/com/znsio/sample/e2e/businessLayer/indigo/GiftVoucherBL.java:78]: you could have used a soft assert here

File: src/test/java/com/znsio/sample/e2e/businessLayer/indigo/GiftVoucherBL.java [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+
+        LOGGER.info("Verify total amount is "+totalAmount+" rupees");
+        assertThat(GiftVoucherScreen.get().getTotalAmount()).isEqualTo(this.totalAmount);
+
+        return this;
+    }
+
+    public GiftVoucherBL applyAPromoCode(String promoCode) {
+
+        LOGGER.info("Apply a "+promoCode+" promo code");
```
Comment (anandbagmar) [inline src/test/java/com/znsio/sample/e2e/businessLayer/indigo/GiftVoucherBL.java:68]: you could have logged this message after you have computed the value for "promo"

File: src/test/java/com/znsio/sample/e2e/businessLayer/indigo/GiftVoucherBL.java [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+
+        LOGGER.info("Preview and Personalise the "+quantity+" Gift Vouchers of "+denominations+" rupees ");
+        GiftVoucherScreen.get().selectQuantity(quantity).selectDenomination(denominations)
+                .personaliseGiftVoucher(details)
+                .previewGiftVoucher();
+
+        totalAmount = Integer.parseInt(quantity) * Integer.parseInt(denominations);
+
+        LOGGER.info("Verify total amount is "+totalAmount+" rupees");
+        assertThat(GiftVoucherScreen.get().getTotalAmount()).isEqualTo(this.totalAmount);
```
Comment (anandbagmar) [inline src/test/java/com/znsio/sample/e2e/businessLayer/indigo/GiftVoucherBL.java:61]: description is missing in all the asserts

File: src/test/resources/com/znsio/sample/e2e/features/indigoFlightPromoCode.feature (hunk 2 above)
Comment (AmanDaharwal) [reply src/test/resources/com/znsio/sample/e2e/features/indigoFlightPromoCode.feature:9]: Done

File: src/test/java/com/znsio/sample/e2e/steps/IndigoFlightSteps.java (hunk 3 above)
Comment (AmanDaharwal) [reply src/test/java/com/znsio/sample/e2e/steps/IndigoFlightSteps.java:45]: Done

File: src/test/java/com/znsio/sample/e2e/steps/IndigoFlightSteps.java (hunk 4 above)
Comment (AmanDaharwal) [reply src/test/java/com/znsio/sample/e2e/steps/IndigoFlightSteps.java:32]: Done

File: src/test/java/com/znsio/sample/e2e/screen/indigo/web/PaymentScreenWeb.java (hunk 5 above)
Comment (AmanDaharwal) [reply src/test/java/com/znsio/sample/e2e/screen/indigo/web/PaymentScreenWeb.java:29]: comment to add - // Amount is in this format "INR 10000.00" - Done

File: src/test/java/com/znsio/sample/e2e/screen/indigo/web/GiftVoucherScreenWeb.java (hunk 6 above)
Comment (AmanDaharwal) [reply src/test/java/com/znsio/sample/e2e/screen/indigo/web/GiftVoucherScreenWeb.java:92]: Done

File: src/test/java/com/znsio/sample/e2e/screen/indigo/web/GiftVoucherScreenWeb.java (hunk 7 above)
Comment (AmanDaharwal) [reply src/test/java/com/znsio/sample/e2e/screen/indigo/web/GiftVoucherScreenWeb.java:17]: added visual validations - done

File: src/test/java/com/znsio/sample/e2e/businessLayer/indigo/GiftVoucherBL.java (hunk 8 above)
Comment (AmanDaharwal) [reply src/test/java/com/znsio/sample/e2e/businessLayer/indigo/GiftVoucherBL.java:102]: Done

File: src/test/java/com/znsio/sample/e2e/businessLayer/indigo/GiftVoucherBL.java (hunk 9 above)
Comment (AmanDaharwal) [reply src/test/java/com/znsio/sample/e2e/businessLayer/indigo/GiftVoucherBL.java:78]: Done

File: src/test/java/com/znsio/sample/e2e/businessLayer/indigo/GiftVoucherBL.java (hunk 10 above)
Comment (AmanDaharwal) [reply src/test/java/com/znsio/sample/e2e/businessLayer/indigo/GiftVoucherBL.java:68]: Done

File: src/test/java/com/znsio/sample/e2e/businessLayer/indigo/GiftVoucherBL.java (hunk 11 above)
Comment (AmanDaharwal) [reply src/test/java/com/znsio/sample/e2e/businessLayer/indigo/GiftVoucherBL.java:61]: Done

File: src/test/resources/reportportal.properties [hunk 12]
```diff
@@ -1,11 +1,11 @@
...
-rp.launch=teswiz
-rp.project=teswiz
-rp.username=default
-rp.enable=false
+rp.endpoint = http://reportportal.jio.com:8080
+rp.uuid = f8379b19-8b89-493e-b5bd-6ffc33096850
+rp.launch = teswiz_TEST_EXAMPLE
+rp.project = teswiz
+rp.enable = true
 rp.description=End-2-End scenarios
```
Comment (raghavgarg1996) [inline src/test/resources/reportportal.properties:6]: No need to push reportportal.jio end points

File: src/test/java/com/znsio/sample/e2e/businessLayer/indigo/GiftVoucherBL.java [hunk 13]
```diff
@@ -0,0 +1,107 @@
...
+    public GiftVoucherBL previewAndPersonaliseGiftVoucher(String quantity, String denominations) {
+
+        LOGGER.info("Navigate to Gift Vouchers");
+        HomeScreen.get().navigateToGiftVouchers();
+
+        LOGGER.info("Preview and Personalise the "+quantity+" Gift Vouchers of "+denominations+" rupees ");
+        GiftVoucherScreen.get().selectQuantity(quantity).selectDenomination(denominations)
+                .personaliseGiftVoucher(details)
+                .previewGiftVoucher();
+
```
Comment (raghavgarg1996) [inline src/test/java/com/znsio/sample/e2e/businessLayer/indigo/GiftVoucherBL.java:57]: Do not use more than one get function in a single function

File: src/test/java/com/znsio/sample/e2e/businessLayer/indigo/GiftVoucherBL.java [hunk 14]
```diff
@@ -0,0 +1,107 @@
...
+        LOGGER.info("Apply a "+promoCode+" promo code");
+        GiftVoucherScreen.get().applyPromoCode(promo);
+
+        softly.assertThat(GiftVoucherScreen.get().getPromoCodeMessage())
+                .as("Promo code message should be equal to expected message")
+                .isEqualTo((String) testData.get("invalidMessage"));
+
+        assertThat(GiftVoucherScreen.get().getTotalAmount())
+                .as("Total amount should be equal to expected amount after applying promo code")
+                .isEqualTo(totalAmount);
```
Comment (raghavgarg1996) [inline src/test/java/com/znsio/sample/e2e/businessLayer/indigo/GiftVoucherBL.java:84]: One single get should be use in a function

File: src/test/java/com/znsio/sample/e2e/screen/indigo/android/GiftVoucherScreenAndroid.java [hunk 15]
```diff
@@ -0,0 +1,170 @@
...
+        this.visually = visually;
+    }
+
+    @Override
+    public GiftVoucherScreen selectQuantity(String quantity) {
+
+        LOGGER.info("select "+quantity+" quantities in gift voucher screen");
+        driver.waitTillElementIsPresent(byQuantityDropdownXpath).click();
+
+        driver.waitTillElementIsPresent(By.xpath("//android.widget.CheckedTextView[contains(@text,'"+quantity+"')]"))
```
Comment (raghavgarg1996) [inline src/test/java/com/znsio/sample/e2e/screen/indigo/android/GiftVoucherScreenAndroid.java:64]: Please do not xpath here

File: src/test/java/com/znsio/sample/e2e/screen/indigo/android/GiftVoucherScreenAndroid.java [hunk 16]
```diff
@@ -0,0 +1,170 @@
...
+        return this;
+    }
+
+    @Override
+    public GiftVoucherScreen selectDenomination(String denominations) {
+
+        LOGGER.info("select "+denominations+" denominations in gift voucher screen");
+        driver.waitTillElementIsPresent(byDenominationDropdownXpath).click();
+
+        driver.waitTillElementIsPresent(By.xpath("//android.widget.CheckedTextView[contains(@text,'"+denominations+"')]"))
```
Comment (raghavgarg1996) [inline src/test/java/com/znsio/sample/e2e/screen/indigo/android/GiftVoucherScreenAndroid.java:76]: Do not use xpath here

File: src/test/java/com/znsio/sample/e2e/screen/indigo/android/GiftVoucherScreenAndroid.java [hunk 17]
```diff
@@ -0,0 +1,170 @@
...
+    }
+
+    @Override
+    public GiftVoucherScreen selectDeliveryOption(String option) {
+        driver.scrollDownByScreenSize();
+
+        LOGGER.info("Select delivery option - "+option);
+
+        if(option.equals("sender")){
+            driver.waitTillElementIsPresent(bySenderXpath).click();
```
Comment (raghavgarg1996) [inline src/test/java/com/znsio/sample/e2e/screen/indigo/android/GiftVoucherScreenAndroid.java:137]: SCreens can not use if/else logic 

Please use the same in Business logic

File: src/test/java/com/znsio/sample/e2e/businessLayer/indigo/GiftVoucherBL.java (hunk 13 above)
Comment (AmanDaharwal) [reply src/test/java/com/znsio/sample/e2e/businessLayer/indigo/GiftVoucherBL.java:57]: Done

File: src/test/java/com/znsio/sample/e2e/businessLayer/indigo/GiftVoucherBL.java (hunk 14 above)
Comment (AmanDaharwal) [reply src/test/java/com/znsio/sample/e2e/businessLayer/indigo/GiftVoucherBL.java:84]: Done

File: src/test/java/com/znsio/sample/e2e/screen/indigo/android/GiftVoucherScreenAndroid.java (hunk 15 above)
Comment (AmanDaharwal) [reply src/test/java/com/znsio/sample/e2e/screen/indigo/android/GiftVoucherScreenAndroid.java:64]: Done

**Pull request closed without merging**